### ISSUE TITLE
Pausing the job queue

### ISSUE THREAD

**Issue by 0xABAB** [distribworks/dkron#240]
Is there a way to temporarily disallow new job submissions?

**Comment by victorcoder** [comment]
Sure, you can set the disabled property of the job to true.

**Comment by 0xABAB** [comment]
@victorcoder I meant a particular mode of the system is which no new jobs will be executed. E.g. suitable for when one wants to upgrade dkron.

**Comment by victorcoder** [comment]
You can call `/leave` to force a node to leave the cluster, then upgrade the node.

Is that what you want?

**Comment by 0xABAB** [comment]
No, semantically that is different. I am talking about an API call to tell the cluster to stop accepting any new jobs, but finish all the currently running ones. 

If all nodes are equal, for upgrading `/leave` would work.

**Comment by kevana** [comment]
I'm not sure I understand the behavior you're looking for. Even if no new jobs are submitted to the cluster, it's cron. All existing jobs would continue to be started at their regularly scheduled times. Setting all jobs to disabled would bring job execution in the cluster to a complete stop, but upgrading nodes one at a time after removing them from the cluster with `/leave` seems like a better way to handle upgrades.

**Comment by 0xABAB** [comment]
@kevana If you set all existing jobs to disabled, people can still define new jobs and for those jobs nothing would be disabled.

**Comment by kevana** [comment]
@0xABAB Understood, to completely bring the world to a stop you'd need to disable all existing jobs and make sure that new jobs are either created in a disabled state or not allowed to be created at all. 

I'd still favor a rolling update strategy. It would be odd to build a piece of infrastructure software that's self-healing and resilient to hardware/network failures but requires the entire cluster to be brought down for upgrades. 100% uptime is the goal, and that goal should apply to the whole lifecycle of the system, including upgrades and scaling.

**Comment by victorcoder** [comment]
I think I see your point here, the title is a bit confusing, a mode where the API stop accepting any new job submission or just completely stop the scheduler, it could be implemented both ways.

Currently there's no way of doing this.

As @kevana say, the goal is not having to stop the job executions, never, though it could be interesting in some use cases.

@0xABAB What do you miss in a rolling update strategy?

**Comment by 0xABAB** [comment]
@kevana I am very much interested in something which can realististicallly consistently get 99.99% availability. What's your relation to this project and perhaps you can comment on #242?

It can be useful to stop a service, if it is under attack for example (run away client process submitting thousands of jobs per second). It's a rather specific use case, and not at all as important as fixing the existing known bugs.

**Comment by kevana** [comment]
It's definitely worthwhile to explore use cases like that, operational issues get really tricky. 

I've been watching this project for a while, distributed systems are a hobby interest of mine. I don't personally run dkron at work.

**Comment by krazykelv** [comment]
This is a feature we need as well.  This can be useful if the jobs are causing a certain part of the system to become overloaded, and we temporarily need to disable them all at once.

For now, a workaround we've found is to create a script to update the type field on each node to something like "under-maintenance".  When our jobs are run against a particular type tag and no nodes match, then the jobs are not scheduled, but existing jobs continue to run.
